ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs

```diff
@@ -0,0 +1,364 @@
+
+        let returnType = skipFunctionParameters parameters fcsType
+        returnType
+
+    let getFunctionParameterTypes parameters (fcsType: FSharpType) =
+        let result = Array.zeroCreate parameters
+        let mutable fullType = fcsType
+
+        for i = 0 to parameters - 1 do
+            result[i] <- fullType.GenericArguments[0]
+            fullType <- fullType.GenericArguments[1]
+
+        result
+
+    // methods take "this" arg as first parameter so skip it
+    let getMethodParameterTypes parameters (fcsType: FSharpType) =
+        let result = Array.zeroCreate parameters
+        let mutable fullType = fcsType.GenericArguments[1]
+
+        for i = 0 to parameters - 1 do
+            result[i] <- fullType.GenericArguments[0]
+            fullType <- fullType.GenericArguments[1]
+
+        result
+
+    let getFunctionParameterAt parameterIndex (fcsType: FSharpType) =
+        let rec getParameter index (fcsType: FSharpType) =
+            if index = 0 then fcsType.GenericArguments[0]
+            else
+                getParameter (index - 1) fcsType.GenericArguments[1]
+
+        getParameter parameterIndex fcsType
+
+module AnnotationUtil =
+
+    let private isFullyAnnotatedNamedTypeUsage (namedTypeUsage: INamedTypeUsage) =
+        isNull namedTypeUsage.ReferenceName ||
+        isNull namedTypeUsage.ReferenceName.TypeArgumentList ||
+        namedTypeUsage.ReferenceName.TypeArgumentList.TypeUsagesEnumerable
+           |> Seq.exists (fun typeUsage -> typeUsage :? IAnonTypeUsage)
+           |> not
+
+    let rec private isFullyAnnotatedFunctionTypeUsage (functionTypeUsage: IFunctionTypeUsage) =
+        isFullyAnnotatedTypeUsage functionTypeUsage.ArgumentTypeUsage &&
+        isFullyAnnotatedTypeUsage functionTypeUsage.ReturnTypeUsage
+
+    and isFullyAnnotatedTypeUsage (typeUsage: ITypeUsage) =
+        match typeUsage with
+        | :? INamedTypeUsage as namedTypeUsage ->
+            isFullyAnnotatedNamedTypeUsage namedTypeUsage
+        | :? IFunctionTypeUsage as functionTypeUsage ->
+            isFullyAnnotatedFunctionTypeUsage functionTypeUsage
+        | :? IConstrainedTypeUsage as constrainedTypeUsage ->
+            isFullyAnnotatedTypeUsage constrainedTypeUsage.TypeUsage
+        | _ -> false
+
+    let isFullyAnnotatedReturnTypeInfo (returnTypeInfo: IReturnTypeInfo) =
+        isNotNull returnTypeInfo.ReturnType &&
+        isFullyAnnotatedTypeUsage returnTypeInfo.ReturnType
+
+    let rec isFullyAnnotatedPattern (pattern: IFSharpPattern) =
+        match pattern.IgnoreInnerParens() with
+        | :? IUnitPat ->
+            true
+        | :? ITypedPat as typedPat ->
+            isFullyAnnotatedPattern typedPat.Pattern
+        | pattern ->
+            pattern |> PatUtil.getTypeUsage |> ValueOption.exists isFullyAnnotatedTypeUsage
+            || pattern |> PatUtil.getReturnTypeInfo |> ValueOption.exists isFullyAnnotatedReturnTypeInfo
```

Comment:
Getting next sibling may not work here, after skipping parens and getting an inner type usage above.

Reply:
Imagine a case like this:
```fsharp
let (Some x): _ option = None
```
If you start with the paren pattern and get its inner pattern afterwards, it won't be able to get the next sibling of the paren pattern. So maybe you should use the inner parent for checking if it's annotated and the outer for getting the next sibling. But, anyway, getting the next sibling manually is not as good as using the generated return type accessor on `IBinding`.

Reply:
I'm getting this from BindingNavigator so it should be safe I guess